### ISSUE TITLE
Wrong shape of hidden_states and attentions when generating

### ISSUE THREAD

**Issue by wywyWang** [huggingface/transformers#26174]
### System Info

- `transformers` version: 4.33.0
- Platform: Windows-10-10.0.22621-SP0
- Python version: 3.10.12
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.3
- Accelerate version: 0.22.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.1 (False)

### Who can help?

@stevhliu, @MKhalusova, @gante

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I would like to get the attentions of the generated tokens using llama. The document of [[GreedySearchDecoderOnlyOutput](https://huggingface.co/docs/transformers/v4.33.0/en/internal/generation_utils#transformers.generation.GreedySearchDecoderOnlyOutput)](https://huggingface.co/docs/transformers/internal/generation_utils#transformers.generation.SampleDecoderOnlyOutput) says that the size of attentions is (num_return_sequences, batch_size, num_heads, **generated_length**, sequence_length) and the size of hidden states is (num_return_sequences, batch_size, **generated_length**, hidden_size) (ignore the tuples for simplicity).  Here the `sequence_length` is the input length of prompt. I thought it might be input_len + generated_len, but the size I printed only contained input_len.

I used the provided example of llama but the size of attention is (num_return_sequences, batch_size, num_heads, **sequence_length**, sequence_length) and the size of hidden states is (num_return_sequences, batch_size, **sequence_length**, hidden_size). I also tried GPT-2 but also got the same behaviors. I am not sure if I missed anything to get the expected results. Thank you.

```python
from transformers import LlamaForCausalLM, LlamaTokenizer
path = "src/llama-2-7b-chat-hf"
tokenizer = LlamaTokenizer.from_pretrained(path)
model = LlamaForCausalLM.from_pretrained(path)
prompt = "Hey, are you conscious? Can you talk to me?"
inputs = tokenizer.encode(prompt, return_tensors="pt")

output_sequences = model.generate(input_ids=inputs, max_length=30, output_attentions=True, output_hidden_states=True, return_dict_in_generate=True)

print(output_sequences.attentions)
print(output_sequences.hidden_states)
```

### Expected behavior

Attention:  (num_return_sequences, batch_size, num_heads, **generated_length**, sequence_length)
Hidden states: (num_return_sequences, batch_size, **generated_length**, hidden_size)

**Comment by Yu-Doit** [comment]
Hi, I'm confused with this behavior too. Do you have any idea about it now? Thanks!

**Comment by gante** [comment]
Hey @wywyWang 👋 

I confirm everything written in the comment above :)

**Comment by sunxiaojie99** [comment]
> Hey @wywyWang, I'm not sure the implementation is wrong. The documentation says the following for `attentions`:
> 
> > Tuple (one element for each generated token) of tuples (one element for each layer of the decoder) of torch.FloatTensor of shape (batch_size, num_heads, generated_length, sequence_length).
> 
> If we print the intermediary values, this is what we get:
> 
> ```python
> from transformers import AutoModelForCausalLM, AutoTokenizer
> path = "gpt2"
> tokenizer = AutoTokenizer.from_pretrained(path)
> model = AutoModelForCausalLM.from_pretrained(path)
> prompt = "Hey, are you conscious?"
> inputs = tokenizer.encode(prompt, return_tensors="pt")
> 
> output_sequences = model.generate(input_ids=inputs, max_length=30, output_attentions=True, output_hidden_states=True, return_dict_in_generate=True)
> 
> for generated_token_index, attention in enumerate(output_sequences.attentions):
>     for i, decoder_element in enumerate(attention):
>         print(f"Generated token index: {generated_token_index}, decoder element {i} shape: {decoder_element.shape}")
> ```
> 
> ```
> Generated token index: 0, shape: torch.Size([1, 12, 6, 6])
> Generated token index: 1, shape: torch.Size([1, 12, 1, 7])
> Generated token index: 2, shape: torch.Size([1, 12, 1, 8])
> Generated token index: 3, shape: torch.Size([1, 12, 1, 9])
> Generated token index: 4, shape: torch.Size([1, 12, 1, 10])
> Generated token index: 5, shape: torch.Size([1, 12, 1, 11])
> Generated token index: 6, shape: torch.Size([1, 12, 1, 12])
> Generated token index: 7, shape: torch.Size([1, 12, 1, 13])
> Generated token index: 8, shape: torch.Size([1, 12, 1, 14])
> Generated token index: 9, shape: torch.Size([1, 12, 1, 15])
> Generated token index: 10, shape: torch.Size([1, 12, 1, 16])
> Generated token index: 11, shape: torch.Size([1, 12, 1, 17])
> Generated token index: 12, shape: torch.Size([1, 12, 1, 18])
> Generated token index: 13, shape: torch.Size([1, 12, 1, 19])
> Generated token index: 14, shape: torch.Size([1, 12, 1, 20])
> Generated token index: 15, shape: torch.Size([1, 12, 1, 21])
> Generated token index: 16, shape: torch.Size([1, 12, 1, 22])
> Generated token index: 17, shape: torch.Size([1, 12, 1, 23])
> Generated token index: 18, shape: torch.Size([1, 12, 1, 24])
> Generated token index: 19, shape: torch.Size([1, 12, 1, 25])
> Generated token index: 20, shape: torch.Size([1, 12, 1, 26])
> Generated token index: 21, shape: torch.Size([1, 12, 1, 27])
> Generated token index: 22, shape: torch.Size([1, 12, 1, 28])
> Generated token index: 23, shape: torch.Size([1, 12, 1, 29])
> ```
> 
> Here we see that:
> 
> * `batch_size` is equal to 1 all the time, which seems good
> * `num_heads` is equal to 12 all the time, which also seems good
> * `generated_length` starts with 6 because that is the initial sequence passed through. It then drops to 1 as the model generates one token at a time, re-using past key values for the previous tokens.
> * `sequence_length` starts with 6, and then continues to update itself up until when it reaches the end of the generation.
> 
> Out of curiosity I passed an explicit `use_cache=False` value to the `generate` method so that it does not leverage past key-values caching, therefore having the attention attend to all tokens at each forward pass:
> 
> ```python
> from transformers import AutoModelForCausalLM, AutoTokenizer
> path = "gpt2"
> tokenizer = AutoTokenizer.from_pretrained(path)
> model = AutoModelForCausalLM.from_pretrained(path)
> prompt = "Hey, are you conscious?"
> inputs = tokenizer.encode(prompt, return_tensors="pt")
> 
> output_sequences = model.generate(input_ids=inputs, max_length=30, output_attentions=True, output_hidden_states=True, return_dict_in_generate=True, use_cache=False)
> 
> for generated_token_index, attention in enumerate(output_sequences.attentions):
>     decoder_element = attention[0]
>     print(f"Generated token index: {generated_token_index}, shape: {decoder_element.shape}")
> ```
> 
> ```
> Generated token index: 0, shape: torch.Size([1, 12, 6, 6])
> Generated token index: 1, shape: torch.Size([1, 12, 7, 7])
> Generated token index: 2, shape: torch.Size([1, 12, 8, 8])
> Generated token index: 3, shape: torch.Size([1, 12, 9, 9])
> Generated token index: 4, shape: torch.Size([1, 12, 10, 10])
> Generated token index: 5, shape: torch.Size([1, 12, 11, 11])
> Generated token index: 6, shape: torch.Size([1, 12, 12, 12])
> Generated token index: 7, shape: torch.Size([1, 12, 13, 13])
> Generated token index: 8, shape: torch.Size([1, 12, 14, 14])
> Generated token index: 9, shape: torch.Size([1, 12, 15, 15])
> Generated token index: 10, shape: torch.Size([1, 12, 16, 16])
> Generated token index: 11, shape: torch.Size([1, 12, 17, 17])
> Generated token index: 12, shape: torch.Size([1, 12, 18, 18])
> Generated token index: 13, shape: torch.Size([1, 12, 19, 19])
> Generated token index: 14, shape: torch.Size([1, 12, 20, 20])
> Generated token index: 15, shape: torch.Size([1, 12, 21, 21])
> Generated token index: 16, shape: torch.Size([1, 12, 22, 22])
> Generated token index: 17, shape: torch.Size([1, 12, 23, 23])
> Generated token index: 18, shape: torch.Size([1, 12, 24, 24])
> Generated token index: 19, shape: torch.Size([1, 12, 25, 25])
> Generated token index: 20, shape: torch.Size([1, 12, 26, 26])
> Generated token index: 21, shape: torch.Size([1, 12, 27, 27])
> Generated token index: 22, shape: torch.Size([1, 12, 28, 28])
> Generated token index: 23, shape: torch.Size([1, 12, 29, 29])
> ```
> 
> We do verify that it does indeed return have a `generated_length` corresponding to all tokens generated without past leveraging of the past key values.
> 
> I'll let @gante comment on whether this is correct or not when back from leave, but the implementation does not seem to be wrong to me.

@LysandreJik Hello! I have a question regarding your example. If I aim to retrieve the last layer hidden states for all tokens, accounting for a total of 30 tokens (7 input tokens and 23 output tokens), it seems that the number of accessible hidden states is 29. This total comprises 7 hidden states from the initial generated token and 22 from the subsequent generated tokens. Is it not possible to obtain the hidden state for the final token, possibly the eos_token?

like this:
```
from transformers import MistralForCausalLM, AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained('mistralai/Mistral-7B-v0.1')
model = MistralForCausalLM.from_pretrained('mistralai/Mistral-7B-v0.1')
prompt = "Hey, are you conscious?"
inputs = tokenizer.encode(prompt, return_tensors="pt")  # shape=[1, 7]

output_sequences = model.generate(input_ids=inputs, max_length=30, output_attentions=True, output_hidden_states=True, return_dict_in_generate=True)

input_hidden = output_sequences.hidden_states[0][-1]  # hidden states of input prompt
output_hidden = torch.cat([item[-1].unsqueeze(0) for item in output_sequences.hidden_states[1:]], dim=0)  # hidden states of generate text

print(input_hidden.shape)  # [bs, 7, dim]
print(output_hidden.shape)  # [22, bs, 1, dim]
```

Expected behavior:
input_hidden: [bs, 7, dim]
output_hidden shape: [23, bs, 1, dim]

**Comment by LukeLIN-web** [comment]
I wonder whether language_model_output.hidden_states[0][-1]. is the first generated token hidden state? How to get the first generated token hidden state?

**Comment by LukeLIN-web** [comment]
> > I wonder whether `language_model_output.hidden_states[0][-1]` is the first generated token hidden state? How to get the first generated token hidden state?
> > From this, we may need generate one more token? [#30036 (comment)](https://github.com/huggingface/transformers/issues/30036#issuecomment-2056766856)
> > Or we just need `outputs.hidden_states[0][-1][:,-1,:]`
> 
> any progress？

I think we can use ` outputs.hidden_states[0][-1][:,-1,:] `

**Comment by ImmortalSdm** [comment]
I'm wondering how we can recover the last attention weights based in cache mode? Any ideas?
